internal/timex/timextest/fake_clock.go

```diff
@@ -0,0 +1,168 @@
+		// called indicates if the timer has already count
+		called bool
+		// index of the timer in the parent fake time source
+		index int
+	}
+)
+
+// NewFakeClock returns a FakeClock with the current time set to the zero value of [time.Time].
+func NewFakeClock() *FakeClock {
+	return &FakeClock{
+		now: time.Time{},
+	}
+}
+
+// Now return the current time.
+func (c *FakeClock) Now() time.Time {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	return c.now
+}
+
+// AfterFunc return a timer that will fire after the specified duration. It is important to note that the clock is
+// locked while the callback is called. This means that you must be cautious about calling any methods on the clock from
+// within the callback. Doing so will probably result in a deadlock. To avoid this, you may want to wrap all such calls
+// in a goroutine.
+func (c *FakeClock) AfterFunc(d time.Duration, f func()) timex.Timer {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	t := &fakeTimer{clock: c, deadline: c.now.Add(d), callback: f}
+	t.index = len(c.timers)
+	c.timers = append(c.timers, t)
+
+	return t
+}
+
+// Update the fake current time.
+func (c *FakeClock) Update(now time.Time) *FakeClock {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	return c.update(now)
+}
+
+// Advance the timer by the specified duration.
+func (c *FakeClock) Advance(d time.Duration) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.update(c.now.Add(d))
+}
+
+// update the fake current time without locking.
+func (c *FakeClock) update(now time.Time) *FakeClock {
+	c.now = now
+	c.fireTimers()
+
+	return c
+}
+
+// fireTimers fires all timers that are ready.
+func (c *FakeClock) fireTimers() {
+	for i := len(c.timers) - 1; i >= 0; i-- {
+		if c.timers[i].deadline.After(c.now) {
+			continue
+		}
+
+		c.timers[i].callback()
+		c.removeTimer(c.timers[i])
+	}
+}
+
+// removeTimer removes the timer from the list of timers. It does so by swapping the timer with the last timer in the
+// list and then shrinking the list by one.
+func (c *FakeClock) removeTimer(f *fakeTimer) bool {
+	if f.called {
+		return false
+	}
+
+	f.called = true
+	c.timers[f.index] = c.timers[len(c.timers)-1] // swap with last timer
+	c.timers[f.index].index = f.index             // update index of swapped timer
+	c.timers = c.timers[:len(c.timers)-1]         // shrink list
+
+	return true
+}
+
+// resetTimer resets the timer to fire after the specified duration.
+func (c *FakeClock) resetTimer(t *fakeTimer, d time.Duration) bool {
+	if t.called {
+		return false
+	}
+
+	t.deadline = c.now.Add(d)
+	c.fireTimers()
+
+	return true
+}
+
+// Reset the timer to fire after the specified duration. Returns true if the timer was active.
+func (t *fakeTimer) Reset(d time.Duration) bool {
+	t.clock.mu.Lock()
+	defer t.clock.mu.Unlock()
+
+	return t.clock.resetTimer(t, d)
+}
+
+// Stop the timer. Returns true if the timer was active.
+func (t *fakeTimer) Stop() bool {
+	t.clock.mu.Lock()
+	defer t.clock.mu.Unlock()
+
+	return t.clock.removeTimer(t)
```

Comment:
If you're willing to loop when firing, you can just loop checking for pointer equality for removal

Reply:
I think I'd rather just do the `index` optimization. I don't think it's too much added complexity.

Reply:
The reverse loop and separate removal function are now gone, so it's just the index of those three